velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+  names.reserve(numFields);
+  types.reserve(numFields);
+  for (auto& field : joinCondition_->expr(0)->distinctFields()) {
+    const auto& name = field->field();
+    auto channel = probeType->getChildIdxIfExists(name);
+    if (channel.has_value()) {
+      auto channelValue = channel.value();
+      filterProbeProjections_.emplace_back(channelValue, filterChannel++);
+      names.emplace_back(probeType->nameOf(channelValue));
+      types.emplace_back(probeType->childAt(channelValue));
+      continue;
+    }
+    channel = buildType->getChildIdxIfExists(name);
+    if (channel.has_value()) {
+      auto channelValue = channel.value();
+      filterBuildProjections_.emplace_back(channelValue, filterChannel++);
+      names.emplace_back(buildType->nameOf(channelValue));
+      types.emplace_back(buildType->childAt(channelValue));
+      continue;
+    }
+    VELOX_FAIL(
+        "Join filter field {} not in probe or build input, filter: {}",
+        field->toString(),
+        filter->toString());
+  }
+
+  filterInputType_ = ROW(std::move(names), std::move(types));
+}
+
+RowVectorPtr NestedLoopJoinProbe::getMismatchedOutput(
+    RowVectorPtr data,
+    const SelectivityVector& matched,
+    BufferPtr& unmatchedMapping,
+    const std::vector<IdentityProjection>& projections,
+    const std::vector<IdentityProjection>& nullProjections) {
+  if (matched.isAllSelected()) {
+    return nullptr;
+  }
+
+  auto rawMapping =
+      initializeRowNumberMapping(unmatchedMapping, data->size(), pool());
+  int32_t numUnmatched{0};
+  for (auto i = 0; i < data->size(); ++i) {
+    if (!matched.isValid(i)) {
+      rawMapping[numUnmatched++] = i;
+    }
+  }
+  VELOX_CHECK_GT(numUnmatched, 0);
+
+  auto output =
+      BaseVector::create<RowVector>(outputType_, numUnmatched, pool());
+  projectChildren(output, data, projections, numUnmatched, unmatchedMapping);
+  for (auto projection : nullProjections) {
+    output->childAt(projection.outputChannel) = BaseVector::createNullConstant(
+        outputType_->childAt(projection.outputChannel), output->size(), pool());
+  }
+  return output;
+}
 
+void NestedLoopJoinProbe::finishProbeInput() {
+  VELOX_CHECK_NOT_NULL(input_);
+  input_.reset();
+  buildIndex_ = 0;
+  if (!noMoreInput_) {
+    return;
+  }
+  if (!needsBuildMismatch(joinType_)) {
+    setState(ProbeOperatorState::kFinish);
+    return;
+  }
+  beginBuildMismatch();
+}
+
+void NestedLoopJoinProbe::noMoreInput() {
+  Operator::noMoreInput();
+  if (state_ != ProbeOperatorState::kRunning || input_ != nullptr) {
+    return;
+  }
+  if (!needsBuildMismatch(joinType_)) {
+    setState(ProbeOperatorState::kFinish);
+    return;
+  }
+  beginBuildMismatch();
+}
+
+void NestedLoopJoinProbe::beginBuildMismatch() {
+  VELOX_CHECK(needsBuildMismatch(joinType_));
+
+  std::vector<ContinuePromise> promises;
+  std::vector<std::shared_ptr<Driver>> peers;
+  if (!operatorCtx_->task()->allPeersFinished(
```

Comment:
It looks like we don't need kWaitForPeers state as we don't need to wait here. We can simply pass null future as HashProbe::noMoreInputInternal() does. Then if it is not the last prober, then we could simply set state to finish and return.

Reply:
That' true.
At first, I thought if you need to grab some state from an operator, it needs to be unfinished. Turns out it don't need to be...